controllers/tcproute_controller.go

```diff
@@ -0,0 +1,231 @@
+		).
+		Complete(r)
+}
+
+// TCProuteReconciler reconciles TCPRoute object
+func (r *TCPRouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	tcproute := new(gatewayv1alpha2.TCPRoute)
+	if err := r.Get(ctx, req.NamespacedName, tcproute); err != nil {
+		if errors.IsNotFound(err) {
+			r.log.Info("object enqueued no longer exists, skipping")
+			return ctrl.Result{}, nil
+		}
+		r.log.Info("Error retrieving tcp route", "Err : ", err)
+		return ctrl.Result{}, err
+	}
+
+	isManaged, gateway, err := r.isTCPRouteManaged(ctx, *tcproute)
+	if !isManaged {
+		// TODO: enable orphan checking https://github.com/Kong/blixt/issues/47
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(tcproute, DataPlaneFinalizer) {
+		if tcproute.DeletionTimestamp != nil {
+			// if the finalizer isn't set, AND the object is being deleted then there's
+			// no reason to bother with dataplane configuration for it its already
+			// handled.
+			return ctrl.Result{}, nil
+		}
+		// if the finalizer is not set, and the object is not being deleted, set the
+		// finalizer before we do anything else to ensure we don't lose track of
+		// dataplane configuration.
+		return ctrl.Result{}, setDataPlaneFinalizer(ctx, r.Client, tcproute)
+	}
+
+	// if the TCPRoute is being deleted, remove it from the DataPlane
+	// TODO: enable deletion grace period https://github.com/Kong/blixt/issues/48
+	if tcproute.DeletionTimestamp != nil {
+		return ctrl.Result{}, r.ensureTCPRouteDeletedInDataPlane(ctx, tcproute, gateway)
+	}
+
+	// in all other cases ensure the TCPRoute is configured in the dataplane
+	if err := r.ensureTCPRouteConfiguredInDataPlane(ctx, tcproute, gateway); err != nil {
+		if err.Error() == "endpoints not ready" {
+			r.log.Info("endpoints not yet ready for TCPRoute, retrying", "namespace", tcproute.Namespace, "name", tcproute.Name)
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// isTCPRouteManaged verifies wether a provided TCPRoute is managed by this
+// controller, according to it's Gateway and GatewayClass.
+func (r *TCPRouteReconciler) isTCPRouteManaged(ctx context.Context, tcproute gatewayv1alpha2.TCPRoute) (bool, *gatewayv1beta1.Gateway, error) {
+	log := log.FromContext(ctx)
+
+	var supportedGateways []gatewayv1beta1.Gateway
+
+	//Use the retrieve objects its parent ref to look for the gateway.
+	var parentRefErr error
+	for _, parentRef := range tcproute.Spec.ParentRefs {
+
+		//Build Gateway object to retrieve
+		gw := new(gatewayv1beta1.Gateway)
+
+		ns := tcproute.Namespace
+		if parentRef.Namespace != nil {
+			ns = string(*parentRef.Namespace)
+		}
+
+		//Get Gateway for TCP Route
+		if err := r.Get(ctx, types.NamespacedName{Name: string(parentRef.Name), Namespace: ns}, gw); err != nil {
+			log.Info("Unable to get parent ref gateway for", "tcpRoute", parentRef.Name, "Namespace", ns, "err", err)
```

Comment:
I think we should check if the error is `NotFound`; in that case, keeping iterating to find a suitable gateway is fine. If we encounter another error, maybe we should immediately return it as an unexpected error occurred. WDYT?

Reply:
(keep in mind that a lot of this controller code is copy/paste from the `UDPRoute` controller)

Yep this is a good suggestion, but it will enable a race condition that can cause a stale route so I've fixed that by adding a `Gateway`->`*Route` map func to both the `TCPRoute` and `UDPRoute` controllers to help compensate:

b1be1b75316230e4088b52fea6e5fe44082e5aaf
a6308da1f5e7e0de4cb95d6c6d6a569a76a8740d

And I've refactored both this method, and the method it was copy/pasted from in the `UDPRoute` controller:

f00e818f61908df9b9fa28d875ab29aa6a121c2f

And they no longer bother collecting errors anymore, any error that is not a 404 is considered a total failure to reconcile.